compiler/rustc_middle/src/traits/solve.rs

```diff
@@ -114,7 +114,6 @@ impl MaybeCause {
 #[derive(Debug, PartialEq, Eq, Clone, Copy, Hash, HashStable, TypeFoldable, TypeVisitable)]
 pub struct QueryInput<'tcx, T> {
     pub goal: Goal<'tcx, T>,
```

Comment:
maybe keep the `defining_opaque_types` in the `QueryInput` (as it's unnecessary to canonicalize it again in the output) and use a `trait QueryInput { fn defining_opaque_types(&self) -> &'tcx List<LocalDefId> }`

Reply:
I tried this, it does not spark joy. I then need to make all other things that are ever used in a `Canonical` take `defining_opaque_types`, which is all `TypeOp`s, as well as their corresponding queries that just take a `ParamEnvAnd<'tcx, some_tuple>`. Doing that would require some groundwork bringing all of these under a common struct... which basically means having a second kind of `Canonical` for the non-`solve::QueryInput` queries.

Reply:
see https://github.com/rust-lang/rust/commit/697f167652db232e79a32098f4af853ad4fb3afe for the (very) unfinished work